src/main/java/at/ac/tuwien/kr/alpha/grounder/transformation/impl/StratifiedEvaluation.java

```diff
@@ -0,0 +1,521 @@
+			// if the starting literal is a builtin or external, we don't have instances in working memory,
+			// but just get fixed substitutions and let bindNextAtomInRule do it's magic ;)
+			FixedInterpretationLiteral fixedInterpretationLiteral = (FixedInterpretationLiteral) startingLiteral;
+			for (Substitution fixedSubstitution : fixedInterpretationLiteral.getSubstitutions(partialStartSubstitution)) {
+				LOGGER.debug("calling bindNextAtom, startingLiteral = {}", startingLiteral);
+				retVal.addAll(this.bindNextAtomInRule(rule, rule.getGroundingOrders().orderStartingFrom(startingLiteral), 0, fixedSubstitution));
+			}
+		} else {
+			Predicate predicate = startingLiteral.getPredicate();
+			if (!this.modifiedInLastEvaluationRun.containsKey(predicate)) {
+				return retVal;
+			}
+			Set<Instance> instances;
+			if (!this.workingMemory.contains(predicate)) {
+				LOGGER.debug("No instances for starting literal " + predicate.toString() + " in working memory!");
+			} else {
+				instances = this.modifiedInLastEvaluationRun.get(predicate);
+				for (Instance inst : instances) {
+					LOGGER.trace("Building substitutions for instance: {}", predicate.getName() + inst.toString());
+					partialStartSubstitution = Substitution.unify(startingLiteral, inst, new Substitution());
+					LOGGER.trace("Got partial substitution for instance: {}", partialStartSubstitution);
+					if (partialStartSubstitution == null) {
+						// continue at this point - when substitution is null it means there's no valid unifier, therefore move on
+						continue;
+					}
+					LOGGER.trace("calling bindNextAtom, startingLiteral = {}", startingLiteral);
+					List<Substitution> subs = this.bindNextAtomInRule(rule, rule.getGroundingOrders().orderStartingFrom(startingLiteral), 0,
+							partialStartSubstitution);
+					retVal.addAll(subs);
+				}
+			}
+		}
+		return retVal;
+	}
+
+	private boolean canFire(InternalRule rule, Substitution substitution) {
+		List<Literal> body = rule.getBody();
+		boolean bodyFulfilled = true;
+
+		for (Literal lit : body) {
+			bodyFulfilled &= this.isLiteralTrue(lit, substitution);
+		}
+		return bodyFulfilled;
+	}
+
+	private void fireRule(InternalRule rule, Substitution substitution) {
+		Atom newAtom = rule.getHeadAtom().substitute(substitution);
+		if (!newAtom.isGround()) {
+			throw new IllegalStateException("Trying to fire rule " + rule.toString() + " with incompatible substitution " + substitution.toString());
+		}
+		LOGGER.debug("Firing rule - got head atom: {}", newAtom);
+		this.workingMemory.addInstance(newAtom, true);
+	}
+
+	private List<Substitution> bindNextAtomInRule(InternalRule rule, RuleGroundingOrder groundingOrder, int orderPosition, Substitution partialSubstitution) {
+		LOGGER.debug("Starting bindNextAtom(\n\trule = {},\n\tgroundingOrder = {},\n\torderPosition = {},\n\tsubstitution = {})", rule,
+				StringUtils.join(groundingOrder, ", "), orderPosition, partialSubstitution);
```

Comment:
Again similar to the logging comments above, but here we should wrap it inside an if-check as follows:
```
		if (LOGGER.isDebugEnabled()) {
			LOGGER.debug("Starting bindNextAtom(\n\trule = {},\n\tgroundingOrder = {},\n\torderPosition = {},\n\tsubstitution = {})", rule,
				StringUtils.join(groundingOrder, ", "), orderPosition, partialSubstitution);
		}
```

Reply:
Fixed (following a larger refactoring after PR #256 got merged)